More antimeridian ("dateline") crossing fixes

gshiroma opened the pull request:
This PR addresses an issue reported by the PGE (thank you, @collinss-jpl) that is affecting some datasets that cross the dateline. The problem is that a buffer declared in the line [here](https://github.com/nasa/PROTEUS/blob/8b6aad1a67f2f4946aae7b5ff046de9df5b66a69/src/proteus/dswx_hls.py#L4385) that is used to check the ancillary input's coverage at the right side of the dateline is too small. In this PR, I'm increasing that buffer from 1 arcsec (0.00027777 deg ~= 30m) to 33 arcsec (about 1km).

I'm taking advantage that we are updating proteus and I'm also adding some other changes:
- Small indentation fixes, white spaces, etc;
- Add condition `min_x < -170` to detect when dateline crossing requires cropping/reprojected of the ancillary input. Dateline crossing only requires reprojection of the ancillary input if this input is in geographic coordinates and its longitude domain is represented using the longitude range [-180, +180] rather than [0, +360]. For example, if a DEM was provided using [0, +360] longitude values, a tile that crosses the dateline (~+180 deg long.) would be fully covered by the DEM without need of cropping or reprojecting it. To detect if the DEM uses the [-180, +180], we just assert that `min_x` is below `-170`.

nemo794 commented on src/proteus/dswx_hls.py:
```diff
@@ -3224,11 +3233,17 @@ def _warp(input_file, geotransform, projection,
-    if not (file_srs.IsGeographic() and tile_min_x < 180 and
-            tile_max_x >= 180):
-
+    # The dateline crossing by the tile is tested by the condition:
+    #     `tile_min_x < 180 and tile_max_x >= 180`
+    # The ancillary input only requires reprojection if it is in
+    # geographic coordinates and if its longitude domain is represented
+    # within the [-180, +180] range, verified by the test `min_x < -170`,
+    # rather than the [0, +360] interval.
+    if not (file_srs.IsGeographic() and file_min_x < -170 and
+            tile_min_x < 180 and tile_max_x >= 180):
+
+        # if it doesn't cross the antimeridian, reproject
+        # input file using gdalwarp
```
Perhaps this could be a bit more straightforward?

```suggestion
    # Flag for if the geotransform (of the e.g. HLS tile) crosses the dateline.
    # The dateline crossing by the tile is tested by the condition:
    #     `tile_min_x < 180 and tile_max_x >= 180`
    flag_tile_crosses_dateline = tile_min_x < 180 and tile_max_x >= 180

    # Flag for if the ancillary input file crosses the dateline.
    # The ancillary input only requires reprojection if it is in
    # geographic coordinates and if its longitude domain is represented
    # within the [-180, +180] range, verified by the test `min_x < -170`,
    # rather than the [0, +360] interval.
    flag_file_crosses_dateline = file_srs.IsGeographic() and file_min_x < -170

    # If neither crosses the antimeridian, reproject input file using gdalwarp
    if not (flag_tile_crosses_dateline and flag_file_crosses_dateline):
```

nemo794 commented on src/proteus/dswx_hls.py:
```diff
@@ -3224,11 +3233,17 @@ def _warp(input_file, geotransform, projection,
+    #     `tile_min_x < 180 and tile_max_x >= 180`
+    # The ancillary input only requires reprojection if it is in
+    # geographic coordinates and if its longitude domain is represented
+    # within the [-180, +180] range, verified by the test `min_x < -170`,
```
What prompted the decision to use -170 ? Could you please add a comment in the code explaining this? Thanks!

gshiroma replied:
There's no specific reason to be `-170`. It could be `-160` or even `0`. The important thing is to know that whether the input is using the longitude range [-180, 180] (but obviously we need some margin to test, we cannot use `-180`).

nemo794 commented on src/proteus/dswx_hls.py:
```diff
@@ -4368,7 +4375,15 @@ def _check_ancillary_inputs(check_ancillary_inputs_coverage,
         flag_error = False
         # Handle antimeridian ("dateline") crossing
-        if file_srs.IsGeographic() and tile_min_x < 180 and tile_max_x >= 180:
+        #
+        # The dateline crossing by the tile is tested by the condition:
+        #     `tile_min_x < 180 and tile_max_x >= 180`
+        # The ancillary input only requires special handling if it is in
+        # geographic coordinates and if its longitude domain is represented
+        # within the [-180, +180] range, verified by the test `min_x < -170`,
+        # rather than the [0, +360] interval.
+        if (file_srs.IsGeographic() and min_x < -170 and
+                tile_min_x < 180 and tile_max_x >= 180):
```
This looks like code duplication. Please pull this logic out into a standalone function. Please have that function use simplified logic akin to the suggestion in my previous comment, and then call that function in both places.

The dateline crossing issue has been a meaningful concern for SDS et al, and thus might have further tweaks in the future. If so, then it would be extremely easy for a maintainer coming in 5 months from now to update the "check" conditions in one location of the code but not the other, but think they'd completed the bug fix.

Thanks!

gshiroma replied:
Thank you, @nemo794 . Please keep in mind that a new function requires docstrings and retesting. We need a quick fix and make sure everything works.

gshiroma replied:
Done

collinss-jpl commented on src/proteus/dswx_hls.py:
```diff
@@ -4382,21 +4425,21 @@ def _check_ancillary_inputs(check_ancillary_inputs_coverage,
 
             # Right side of the antimeridian crossing: +180 -> +360
-            buffer_in_degrees = 0.0002777  # buffer of 1 arcsec: ~ 30m
+            buffer_in_degrees = 33 * 0.0002777  # buffer of 33 arcsec: ~ 1km
```
I think it'll be handy to assign `33` to a constant near the top of the file so its easy to modify going forward, just in case we find another tile that needs a slightly larger buffer. Same for the `0.0002777`.

gshiroma replied:
Just did it. Thank you, @collinss-jpl

collinss-jpl reviewed (approved):
Had one suggestion to assign some magic numbers to constants, but otherwise this looks good to me, thanks for the quick work as always @gshiroma!

nemo794 commented on src/proteus/dswx_hls.py:
```diff
@@ -3136,6 +3136,44 @@ def get_projection_proj4(projection):
+        MGRS tile min longitude value in degrees
+    tile_max_x: float
+        MGRS tile max longitude value in degrees
+
```
Could you please explain what this function returns? Thanks!

nemo794 commented on src/proteus/dswx_hls.py:
```diff
@@ -3155,17 +3155,21 @@ def _antimeridian_crossing_requires_special_handling(
 
+    # Flag to test if the ancillary input file is in geographic
+    # coordinates and if its longitude domain is represented
+    # within the [-180, +180] range, rather than the [0, +360] interval.
+    # This is verified by the test `min_x < -170`,
```
Suggestion:
```
    # This is verified by the test `min_x < -170`. There's no specific reason why
    # -170 is used. It could be -160, or even 0.
 ```

gshiroma commented:
> Had one suggestion to assign some magic numbers to constants, but otherwise this looks good to me, thanks for the quick work as always @gshiroma!

You provided me with all the info I needed to identify and solve this issue. Thank you so much, @collinss-jpl !

nemo794 reviewed (approved):
Thank you for addressing the comments; these changes should help in the long run. I ran a couple of test granules (one that crosses the dateline, the second one not), and both ran to completion. Thanks, @gshiroma !

gshiroma commented:
> Thank you for addressing the comments; these changes should help in the long run. I ran a couple of test granules (one that crosses the dateline, the second one not), and both ran to completion. Thanks, @gshiroma !

Thank you, @nemo794 , for reviewing the code and testing it over a couple of granules!
